[Disputes] Introduce a loading state to the “Challenge dispute” (new evidence) flow

mgascam opened the pull request:
Fixes #WOOPMNT-5307

#### Changes proposed in this Pull Request

This PR introduces a loading state that helps to fix a race condition in the disputes E2E tests. In the "Save for later" flow, the description is changed from "Beanie" to "my product description" and saved. However, when the dispute is revisited later, the form incorrectly shows the old value ("Beanie") instead of the updated one. [Run logs](https://github.com/Automattic/woocommerce-payments/actions/runs/17491231822/job/49681367080)

```
  1) [chromium] › tests/e2e/specs/wcpay/merchant/merchant-disputes-respond.spec.ts:527:6 › Disputes > Respond to a dispute › Save a dispute challenge without submitting evidence › Select product type and fill description 

    Error: Timed out 20000ms waiting for expect(locator).toHaveValue(expected)

    Locator: getByLabel('PRODUCT DESCRIPTION')
    Expected string: "my product description"
    Received string: "Beanie"
    Call log:
      - expect.toHaveValue with timeout 20000ms
      - waiting for getByLabel('PRODUCT DESCRIPTION')
        24 × locator resolved to <textarea rows="4" id="inspector-textarea-control-0" class="components-textarea-control__input css-1x3y2gc e1w5nnrk0">Beanie</textarea>
           - unexpected value "Beanie"


      589 | 				await expect(
      590 | 					merchantPage.getByLabel( 'PRODUCT DESCRIPTION' )
    > 591 | 				).toHaveValue( 'my product description' );
          | 				  ^
      592 | 			}
      593 | 		);
      594 |
        at /home/runner/work/woocommerce-payments/woocommerce-payments/tests/e2e/specs/wcpay/merchant/merchant-disputes-respond.spec.ts:591:7
        at /home/runner/work/woocommerce-payments/woocommerce-payments/tests/e2e/specs/wcpay/merchant/merchant-disputes-respond.spec.ts:573:3
``` 

Introducing the initial loading state removes the UI race that caused flakiness and also improves the perceived performance and UX flow.

Here's a screen recording of the new spinner in action:


https://github.com/user-attachments/assets/2121f3f9-a921-452f-b2f9-6409a735f911





<!--
Title: A descriptive, yet concise, title.
-->

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

<!--
Questions for PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

#### Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* E2E tests should pass

Manual tests of the flow should work as expected:

1. As as shopper, checkout a product paying with a disputes test card i.e: 4000000000002685
2. As a merchant, navigate to Payments > Disputes and locate the generated dispute. Click it to enter the details.
3. In the details page, click the Challenge dispute button
4. Check that the spinner appears while the dispute details are loaded.
5. In the challenge dispute form, edit the Product Description ie: "My product description"
6. Click Save for later
7. Repeat steps 2-3
8. In the challenge dispute form, ensure the Description shows your edited description i.e. "My product description"

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

mgascam committed:
Version coverage matrix + retry logic + docs

mgascam committed:
Stabilize flaky specs and helpers

mgascam committed:
[Disputes] Persist user-entered values on revisit

botwoo commented:
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `11037` or branch name `dev/woopmnt-5249-disputes-ui` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=dev/woopmnt-5249-disputes-ui) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: 2b08ea40b2d5cb20aa512a78f42575d63f102b5d
- Build time: 2025-09-19 08:25:55 UTC

_Note: the build is updated when a new commit is pushed to this PR._

mgascam committed:
Include disputes UI persistence needed to deflake tests

mgascam commented:
Closing since this change is needed to stabilize flaky tests in https://github.com/Automattic/woocommerce-payments/pull/11036

mgascam committed:
Add changelog

mgascam committed:
Revert "Include disputes UI persistence needed to deflake tests"

This reverts commit 9370ddce731a2d9a6def7a7e8992304964c1b2fb.

mgascam committed:
PR feedback: amend docs

mgascam committed:
Merge branch 'develop' into dev/woopmnt-5249-e2e-version-coverage-and-ci

mgascam committed:
Use npm script for re-running failed specs

mgascam committed:
Merge branch 'dev/woopmnt-5249-e2e-version-coverage-and-ci' into dev/woopmnt-5249-e2e-fixes

mgascam committed:
Skip flaky test

mgascam commented:
Re-opening, see [comment ](https://github.com/Automattic/woocommerce-payments/pull/11036#discussion_r2324806281)

mgascam committed:
Merge branch 'dev/woopmnt-5249-e2e-fixes' into dev/woopmnt-5249-disputes-ui

mgascam committed:
Re-enable skipped test

mgascam committed:
Merge branch 'develop' into dev/woopmnt-5249-e2e-fixes

mgascam committed:
Merge branch 'dev/woopmnt-5249-e2e-fixes' into dev/woopmnt-5249-disputes-ui

mgascam commented:
Hi @rtio it would be great if you could review this PR, please see [comment](https://github.com/Automattic/woocommerce-payments/pull/11036#discussion_r2324271373) for context. Thanks!

rtio reviewed (commented):
I tested the changes and they seem to work fine, but I couldn’t actually reproduce the bad behavior in `develop`. From your comments, it looks like this only happens in E2E tests when `d?.evidence` isn’t fully loaded yet. Adding the extra complexity to handle that across all fields feels like a heavy workaround. A cleaner alternative might be to show a loading state until the dispute is fully loaded. Thoughts?

mgascam commented:
> A cleaner alternative might be to show a loading state until the dispute is fully loaded. Thoughts?

Thanks for the review @rtio, I’ll give the loading state approach a try.

mgascam committed:
Merge branch 'develop' into dev/woopmnt-5249-disputes-ui

mgascam committed:
Merge branch 'develop' into dev/woopmnt-5249-disputes-ui

mgascam committed:
Enhance dispute evidence submission flow with loading state handling and test updates

mgascam committed:
Re-enable skipped test

mgascam committed:
Add changelog

mgascam commented:
Hi @rtio, I added the loading state as you suggested in [this comment](https://github.com/Automattic/woocommerce-payments/pull/11037#pullrequestreview-3190538989), and it’s ready for review.

@lucyneb, it would be great to get your input as well since you were involved in the redesign of the disputes flow. For context: I added a loading state (screen recording attached in the description) to address an issue that surfaced during E2E runs. The tests run so fast that the form state can be unexpected. This wasn’t affecting “real” users, but it’s something we needed to address. In my view, the loading state not only fixes the issue but also improves the UX. Curious to hear your thoughts.

Thanks 🙏

rtio reviewed (approved):
Just tested. Feels like an huge improvement! Thank you for working on that.

mgascam commented:
Hi @htdat, I saw the [pr: needs author reply](https://github.com/Automattic/woocommerce-payments/issues?q=state%3Aopen%20label%3A%22pr%3A%20needs%20author%20reply%22) label was added. I might be overlooking something, but I don’t see any pending feedback. Could you point me to what needs my attention?

htdat commented:
@mgascam - I thought that you're waiting for feedback from Design :) Feel free to merge, I do not have any further comment.

mgascam committed:
Merge branch 'develop' into dev/woopmnt-5249-disputes-ui

mgascam committed:
Merge branch 'develop' into dev/woopmnt-5249-disputes-ui

mgascam committed:
Merge branch 'develop' into dev/woopmnt-5249-disputes-ui

lucyneb commented:
@mgascam looks good! thanks for working on this. The loader you used comes from the design system, right? Just double checking. Otherwise approved on my end.

mgascam committed:
Merge branch 'develop' into dev/woopmnt-5249-disputes-ui

mgascam commented:
> @mgascam looks good! thanks for working on this. The loader you used comes from the design system, right? Just double checking. Otherwise approved on my end.

Thanks for the review @lucyneb. Yes, I confirm I'm using the `Spinner` from the `@wordpress/components` package.
